src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -2714,6 +2725,28 @@ CheckConnectionTimeout(WorkerPool *workerPool)
 }
 
 
+/*
+ * MarkEstablishingSessionsTimedOut goes over the sessions in the given
+ * workerPool and marks them timed out. ConnectionStateMachine()
+ * later cleans up the sessions.
+ */
+static void
+MarkEstablishingSessionsTimedOut(WorkerPool *workerPool)
+{
+	WorkerSession *session = NULL;
+	foreach_ptr(session, workerPool->sessionList)
+	{
+		MultiConnection *connection = session->connection;
+
+		if (connection->connectionState == MULTI_CONNECTION_CONNECTING ||
+			connection->connectionState == MULTI_CONNECTION_INITIAL)
+		{
+			connection->connectionState = MULTI_CONNECTION_TIMED_OUT;
```

Comment:
could one of the connections already be sending a command? does that matter?

Reply:
As far as I can see, that could happen such as:
- `set  citus.force_max_query_parallelization TO ON;` and replication factor == 2
- set max_cached_connections > 1

Still, we are only failing the sessions that has not been able to established (and not started pop'in placement executions via `PopUnassignedPlacementExecution()`). I don't see a problem even if some concurrent sessions has already started sending commands. That is part of the regular execution flow anyway. 

Depending on the state, the overall execution might fail or mark some placements invalid, which should also be OK. Otherwise, the other sessions should be able to rely on `PopUnassignedPlacementExecution()` / `PopAssignedPlacementExecution()` to get new stuff for execution.

Reference for readers/future: The problem before this patch was the timed-out connections were coming up as if they were initiated normally. But, in the window between the connection has timed out and it comes back as connected, the executor already decides to use other connections for certain tasks. Later, the timed out connection tries to execute queries, which leads to various issues.

Reply:
makes sense